File: MatchMe-tvOSGameObjC Cropped/MatchMe/MatchMe/CardDataSource.m
//
//  CardDataSource.m
//  MatchMe
//
//  Created by Siyana Slavova on 2/23/16.
//  Copyright © 2016 mycompany.MatchMe. All rights reserved.
//

#import "CardDataSource.h"

#import "CardCollectionViewCell.h"

#import "NSMutableArray+Shuffle.h"


@interface CardDataSource ()

@property (strong, nonatomic) NSMutableArray *cards;

@property (nonatomic) LevelMode levelMode;

@property (nonatomic, readwrite) NSInteger columnsCount;
@property (nonatomic, readwrite) NSInteger rowsCount;

@end

@implementation CardDataSource

- (instancetype)initWithLevelMode:(LevelMode)levelMode {
    self = [super init];
    if (self) {
        self.levelMode = levelMode;
        [self createCardsArrays];
    }
    return self;
}

- (NSMutableArray *)cards {
    if (!_cards) {
        _cards = [NSMutableArray new];
    }
    return _cards;
}

#pragma mark- CollectionViewDataSource
- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView {
    return  self.rowsCount;
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return self.columnsCount;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    CardCollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"CardCellIdentifier" forIndexPath:indexPath];
    Card *card = [self cardAtIndexPath:indexPath];
    cell.frontImageView.image = card.cardImage;
    return cell;
}

#pragma mark - Help methods

- (NSInteger)uniqueCardsCount {
    // in cards array all cards are doubled
    return self.cards.count/2;
}

- (Card *)cardAtIndexPath:(NSIndexPath *)indexPath {
    NSInteger index = indexPath.row + (indexPath.section * self.columnsCount);
    if (index < 0 || index > self.cards.count) {
        return nil;
    }
    Card *card = self.cards[index];
    NSLog(@"Card index: %ld", (long)index);
    return card;
}

- (void)createCardsArrays {
    // form the images array
    NSMutableArray *finalNames = nil;
    NSInteger cardCount = 0;
    switch (self.levelMode) {
        case LevelModeEasy:
        {
            finalNames = [@[@"appleW", @"blueberriesW", @"carrotW", @"lettuceW", @"mushroomW", @"potatoW", @"radishW", @"strawberryW", @"sugar_snapW", @"tomatoW"] mutableCopy];
            cardCount = easyLevelCardCount;
            self.columnsCount = 4;
            self.rowsCount = 2;
        }
            break;
        case LevelModeNormal:
        {
            finalNames = [@[@"1_fishW", @"antW", @"bear_mac_archigraphsW", @"dragon_flyW", @"elephantW", @"hp_boyW", @"hp_catW", @"hp_dogW", @"hp_girlW", @"ladybugW", @"penguinW", @"silly_boy_archigraphsW", @"turtleW"] mutableCopy];
            cardCount = normalLevelCardCount;
            self.columnsCount = 4;
            self.rowsCount = 3;
        }
            break;
        case LevelModeHard:
        {
            finalNames = [@[@"aeroplaneW", @"bugattiW", @"busW", @"carW", @"suitcaseW", @"trainW", @"travel_busW", @"1_fishW", @"antW", @"bear_mac_archigraphsW", @"dragon_flyW", @"elephantW", @"hp_boyW", @"hp_catW", @"hp_dogW", @"hp_girlW", @"ladybugW", @"penguinW", @"silly_boy_archigraphsW", @"turtleW"] mutableCopy];
            cardCount = hardLevelCardCount;
            self.columnsCount = 6;
            self.rowsCount = 3;
        }
            break;
            
        default:
            break;
    }
    
    [finalNames shuffle];
    
    NSMutableArray *cardPicturesNames = [NSMutableArray new];
    
    // Double the pictures
    for (int i = 0; i < cardCount/2; i++) {
        [cardPicturesNames addObject:finalNames[i]];
        [cardPicturesNames addObject:finalNames[i]];
    }
    
    [cardPicturesNames shuffle];
    
    for (NSString *cardName in cardPicturesNames) {
        Card *card = [[Card alloc] initWithName:cardName];
        [self.cards addObject:card];
    }
    
    
}
@end

File: MatchMe-tvOSGame Cropped/MatchMe/CardDataSource.swift
//
//  CardDataSource.swift
//  MatchMe
//
//  Created on 10/19/15.
//  Copyright © 2015. All rights reserved.
//

import UIKit

class CardDataSource: NSObject {
    
    var cards = [Card]()
    var cardPicturesNames = [String]()
    private var easyLevelImages = [String]()
    private var normalLevelImages = [String]()
    private var hardLevelImages = [String]()

    override init() {
        super.init()
        self.easyLevelImages = ["appleW", "blueberriesW", "carrotW", "lettuceW", "mushroomW", "potatoW", "radishW", "strawberryW", "sugar_snapW", "tomatoW"]
        self.normalLevelImages = ["1_fishW", "antW", "bear_mac_archigraphsW", "dragon_flyW", "elephantW", "hp_boyW", "hp_catW", "hp_dogW", "hp_girlW", "ladybugW", "penguinW", "silly_boy_archigraphsW", "turtleW"]
        self.hardLevelImages = ["aeroplaneW", "bugattiW", "busW", "carW", "suitcaseW", "trainW", "travel_busW"]
        
        let cardCount = levelRowNumber*levelColumnNumber
        var finalNames = [String]()
        
        if (cardCount == easyLevelCardCount) {
            finalNames = self.easyLevelImages.shuffle()
        } else if (cardCount == normalLevelCardCount) {
            finalNames = self.normalLevelImages.shuffle()
        } else if (cardCount == hardLevelCardCount) {
            finalNames = self.normalLevelImages + self.hardLevelImages
            finalNames = finalNames.shuffle()
        }
        
        //Double the pictures
        for nameIndex in 0...(cardCount/2 - 1) {
            self.cardPicturesNames.append(finalNames[nameIndex])
            self.cardPicturesNames.append(finalNames[nameIndex])
        }
        
        //Make names images
        self.cardPicturesNames = self.cardPicturesNames.shuffle()
        for var index = self.cardPicturesNames.count - 1; index >= 0; --index {
            if let cardImage = UIImage(named: self.cardPicturesNames[index]) {
                self.cards.append(Card(name: self.cardPicturesNames[index], image: cardImage))
            }
        }
    }

    func cardPictureAtIndex(index: Int) -> Card? {
        if index < 0 || index > self.cards.count {
            return nil
        } else {
            return self.cards[index]
        }
    }
}

extension CardDataSource: UICollectionViewDataSource {
    func numberOfSectionsInCollectionView(collectionView: UICollectionView) -> Int {
        return levelRowNumber
    }
    
    func collectionView(collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return levelColumnNumber
    }
    
    func collectionView(collectionView: UICollectionView, cellForItemAtIndexPath indexPath: NSIndexPath) -> UICollectionViewCell {
        guard let cell = collectionView.dequeueReusableCellWithReuseIdentifier("CardCellIdentifier", forIndexPath: indexPath) as? CardCollectionViewCell else {
            return UICollectionViewCell()
        }
        let card : Card = self.cardPictureAtIndex(indexPath.row + (levelColumnNumber*indexPath.section))!
        cell.frontImageView.image = card.image
        NSLog ("\(indexPath.row + (levelColumnNumber*indexPath.section))")
        return cell
    }

}

extension CollectionType {
    /// Return a copy of `self` with its elements shuffled
    func shuffle() -> [Generator.Element] {
        var list = Array(self)
        list.shuffleInPlace()
        return list
    }
}

extension MutableCollectionType where Index == Int {
    /// Shuffle the elements of `self` in-place.
    mutating func shuffleInPlace() {
        // empty and single-element collections don't shuffle
        if count < 2 { return }
        
        for i in 0..<count - 1 {
            let j = Int(arc4random_uniform(UInt32(count - i))) + i
            guard i != j else { continue }
            swap(&self[i], &self[j])
        }
    }
}
